File: algorithms/java/Jun.2020.challenge/Challenge_Random_Pick_with_Weight.java
class Solution {
    int[] w;
    Random r;
    public Solution(int[] w) {
        this.w = w;
        r = new Random();
        for (int i = 1; i < w.length; i++) {
            w[i] += w[i - 1];
        }
    }
    
    public int pickIndex() {
        int lastSum = w[w.length - 1];
        int target = r.nextInt(lastSum) + 1;
        int p = Arrays.binarySearch(w, target);
        return p >= 0 ? p : -p - 1;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

File: algorithms/python/Jun.2020.challenge/Challenge_Random_Pick_with_Weight.py
from bisect import bisect_left

class Solution:

    def __init__(self, w: List[int]):
        self.w = w
        for i in range(1, len(w)):
            self.w[i] = self.w[i] + self.w[i-1] 

    def pickIndex(self) -> int:
        idx = random.randint(1, self.w[-1])
        return bisect_left(self.w, idx)

# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()
